File: src/scripts/dbt2-post-process.jl
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright (C) 2006      Open Source Development Labs, Inc.
#               2014      2ndQuadrant, Ltd.
#               2006-2023 Mark Wong
#

#=
if [ $# -lt 1 ]; then
    echo "$(basename "$0") is the DBT-2 mix log analyzer"
    echo ""
    echo "Usage:"
    echo "  $(basename "$0") VERBOSE mix.log [mix-1.log [...]]"
    echo
fi

if [ ! "$VERBOSE" = "1" ]; then
	VERBOSE=0
fi

exec julia --color=no --startup-file=no "$0" $VERBOSE "$@"
=#

using CSV
using DataFrames
using Distributions
using Printf
using Statistics

const TransactionNames = Dict(
    "d" => "Delivery",
    "n" => "New Order",
    "o" => "Order Status",
    "p" => "Payment",
    "s" => "Stock Level"
)

function display_did_p(df, wdf, stats)
    min = minimum(wdf.wid)
    max = maximum(wdf.wid)

    pdf_wid_did = pivot_by_wid_did(df, stats)
    # FIXME: Should determine district ranged based on test parameters as
    #        opposed to hard coding the default10.
    for wid = min:max
        E = wdf[wdf.wid .== wid, :].nrow[1] / 10.0
        chi_square = 0
        for did = 1:10
            O = pdf_wid_did[(pdf_wid_did.wid .== wid) .&
                            (pdf_wid_did.did .== did), :].nrow[1]
            chi_square += (O - E)^2
        end
        chi_square /= E
        p = 1 - cdf(Chisq(9), chi_square)
        @printf("* Warehouse %d District chi-square p-value: %f\n", wid, p)
    end
end

function display_wid_p(wdf, stats)
    min = minimum(wdf.wid)
    max = maximum(wdf.wid)
    total = sum(wdf.nrow)
    E = total / (max - min + 1)
    chi_square = 0
    for wid = min:max
        O = wdf[wdf.wid .== wid, :].nrow[1]
        chi_square += (O - E)^2
    end
    chi_square /= E
    p = 1 - cdf(Chisq(max - min), chi_square)
    @printf("* Warehouse (%d-%d) chi-square p-value: %f\n", min, min, p)
end

function displayreport(summary_txn, summary_txn_code, stats)
    TotalTransactions = sum(summary_txn.nrow)

    @printf("============  =====  =========  =========  ===========  ")
    @printf("===========  =====\n")
    @printf("          ..     ..    Response Time (s)            ..")
    @printf("           ..     ..\n")
    @printf("------------  -----  --------------------  -----------  ")
    @printf("-----------  -----\n")
    @printf(" Transaction      %%   Average     90th %%        Total")
    @printf("    Rollbacks      %%\n")
    @printf("============  =====  =========  =========  ===========  ")
    @printf("===========  =====\n")

    rollbacks = 0
    for i in ["d", "n", "o", "p", "s"]
        row = summary_txn[(summary_txn.transaction .== i), :]
        try
            rollbacks = summary_txn_code[
                    (summary_txn_code.transaction .== i) .&
                    (summary_txn_code.code .== "R"), :].nrow[1]
        catch
            rollbacks = 0
        end

        @printf("%12s  %5.2f  %9.3f   %8.3f  %11d  %11d  %5.2f\n",
                TransactionNames[i],
                row.nrow[1] / TotalTransactions * 100,
                row.avg[1],
                row.q90[1],
                row.nrow[1],
                rollbacks,
                rollbacks / (row.nrow[1]) * 100)
    end
    @printf("============  =====  =========  =========  ===========  ")
    @printf("===========  =====\n")
    @printf("\n")
    @printf("* Throughput: %0.2f new-order transactions per minute (NOTPM)\n",
            summary_txn[(summary_txn.transaction .== "n"), :].nrow[1] /
                    (stats["SteadyStateTime"] / 60))
    @printf("* Duration: %0.1f minute(s)\n", stats["SteadyStateTime"] / 60)
    @printf("* Unknown Errors: %d\n", stats["errors"])
    @printf("* Ramp Up Time: %0.1f minute(s)\n", stats["RampupTime"] / 60)
end

function load(filenames)
    colnames = [
            "ctime",
            "transaction",
            "code",
            "response_time",
            "id",
            "wid",
            "did",
    ]
    df = mapreduce(vcat, filenames) do filename
        DataFrame(CSV.File(filename, header=colnames))
    end
    return df
end

function pivot_by_txn(df, stats)
    gdf = groupby(filter(row -> row["ctime"] > stats["StartTime"] &&
                         row["ctime"] < stats["TerminatedTime"] &&
                         row["code"] != "E", df),
                  [:transaction]; sort=true, skipmissing=true)
    summary = combine(gdf,
                 nrow,
                 :response_time => mean => :avg,
                 :response_time => (x -> quantile(x, .90)) => :q90
    )
    return summary
end

function pivot_by_txn_code(df, stats)
    gdf = groupby(filter(row -> row["ctime"] > stats["StartTime"] &&
                         row["ctime"] < stats["TerminatedTime"], df),
                  [:transaction, :code]; sort=true, skipmissing=true)
    summary = combine(gdf, nrow)
    return summary
end

function pivot_by_wid(df, stats)
    gdf = groupby(filter(row -> row["ctime"] > stats["StartTime"] &&
                         row["ctime"] < stats["TerminatedTime"], df),
                  [:wid]; skipmissing=true)
    summary = combine(gdf, nrow)
    return summary
end

function pivot_by_wid_did(df, stats)
    gdf = groupby(filter(row -> row["ctime"] > stats["StartTime"] &&
                         row["ctime"] < stats["TerminatedTime"], df),
                  [:wid, :did]; skipmissing=true)
    summary = combine(gdf, nrow)
    return summary
end

function prep(df)
    stats = Dict()

    stats["errors"] = count(y -> (y .== "E"), skipmissing(df.code))

    stats["StartTime"] = maximum(df[df.transaction .== "START", :ctime])
    stats["TerminatedTime"] = minimum(df[df.transaction .== "TERMINATED",
                                      :ctime])
    stats["SteadyStateTime"] = stats["TerminatedTime"] - stats["StartTime"]

    stats["ctime0"] = df[[1], [1]].ctime[1]
    stats["RampupTime"] = stats["StartTime"] - stats["ctime0"]

    return stats
end

function main()
    df = load(ARGS[2:end])
    stats = prep(df)
    pdf_txn = pivot_by_txn(df, stats)
    pdf_txn_code = pivot_by_txn_code(df, stats)
    displayreport(pdf_txn, pdf_txn_code, stats)
    if ARGS[1] == "1"
        wdf = pivot_by_wid(df, stats)
        display_wid_p(wdf, stats)
        display_did_p(df, wdf, stats)
    end
end

main()

File: src/scripts/dbt2-post-process.r
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright (C) 2006      Open Source Development Labs, Inc.
#               2014      2ndQuadrant, Ltd.
#               2006-2022 Mark Wong
#

MIXFILE=$1

if [ $# -lt 1 ]; then
	echo "$(basename $0) is the DBT-2 mix log analyzer"
	echo ""
	echo "Usage:"
	echo "  $(basename $0) mix.log [mix-1.log [...]]"
	echo
fi

if [ ! "x$VERBOSE" = "x1" ]; then
	VERBOSE=0
fi

# Covert the list of file names from the comand line into a quoted and comma
# separated list for R.
FILENAMES=""
for FILENAME in $@; do
	FILENAMES="$FILENAMES \"$FILENAME\""
done
FILENAMES=$(echo $FILENAMES | sed -e "s/ /,/g")

R --slave --no-save << __EOF__
filenames <- c($FILENAMES)
mix <- do.call(rbind, lapply(filenames, read.csv, header=FALSE))

colnames(mix)[1] <- 'ctime'
colnames(mix)[2] <- 'txn'
colnames(mix)[3] <- 'status'
colnames(mix)[4] <- 'response'
colnames(mix)[5] <- 'id'
colnames(mix)[6] <- 'wid'
colnames(mix)[7] <- 'did'

# Get the ctime of the last START marker and the ctime the first TERMINATED
# marker to calculate steady state.
start <- max(mix\$ctime[mix\$txn == "START"])
end <- min(mix\$ctime[mix\$txn == "TERMINATED"])

total_txn <- sum(mix\$ctime > start & mix\$ctime < end)

duration = end - start
errors <- sum(mix\$status == "E")

cat(paste("============  =====  =========  =========  ===========  ",
          "===========  =====\n", sep = ""))
cat(paste("          ..     ..    Response Time (s)            ..  ",
          "         ..     ..\n", sep = ""))
cat(paste("------------  -----  --------------------  -----------  ",
          "-----------  -----\n", sep = ""))
cat(paste(" Transaction      %   Average     90th %        Total    ",
          "Rollbacks      %\n", sep = ""))
cat(paste("============  =====  =========  =========  ===========  ",
          "===========  =====\n", sep = ""))

txn <- c("d", "n", "o", "p", "s")
txn_name <- c("Delivery", "New Order", "Order Status", "Payment",
              "Stock Level")

for (i in 1:5) {
  t_total <- sum(mix\$txn == txn[i] & mix\$ctime > start & mix\$ctime < end)
  if (i == 2) {
    total_n <- t_total
  }
  t_rollback <- sum(mix\$txn == txn[i] & mix\$status == "R" &
                    mix\$ctime > start & mix\$ctime < end, na.rm=TRUE)
  t_mean <- mean(mix\$response[mix\$txn == txn[i] & mix\$ctime > start &
                 mix\$ctime < end], na.rm=TRUE)
  t_q90 <- quantile(mix\$response[mix\$txn == txn[i] & mix\$ctime > start &
                    mix\$ctime < end], .9, na.rm=TRUE)
  cat(sprintf("%12s  %5.2f  %9.3f  %9.3f  %11d  %11d  %5.2f\n",
              txn_name[i], t_total / total_txn * 100, t_mean, t_q90,
              t_total, t_rollback, t_rollback / t_total * 100))
}

cat(paste("============  =====  =========  =========  ===========  ",
          "===========  =====\n\n", sep = ""))
cat(sprintf("* Throughput: %0.2f new-order transactions per minute (NOTPM)\n",
            total_n / (duration / 60)))
cat(sprintf("* Duration: %0.1f minute(s)\n", duration / 60))
cat(sprintf("* Unknown Errors: %d\n", errors))

# Calculate the ramp up time.

duration <- start - min(mix\$ctime)
cat(sprintf("* Ramp Up Time: %0.1f minute(s)\n", duration / 60))

if ($VERBOSE == 1) {
  wid_min <- min(mix\$wid, na.rm=TRUE)
  wid_max <- max(mix\$wid, na.rm=TRUE)

  E_w <- total_txn / (wid_max - wid_min + 1)
  chi_square_w <- 0
  for (i in wid_min:wid_max) {
    O_w <- sum(mix\$wid == i & mix\$ctime > start & mix\$ctime < end)
    chi_square_w <- chi_square_w + (O_w - E_w)^2
  }
  chi_square_w <- chi_square_w / E_w
  p_w <- 1 - pchisq(chi_square_w, df=(wid_max - wid_min))
  cat(sprintf("* Warehouse (%d-%d) chi-square p-value: %f\n",
              wid_min, wid_max, p_w))

  for (i in wid_min:wid_max) {
    # FIXME: Should know district range used for test as opposed to hard code
    #        to 10.
    E_d <- O_w / 10
    chi_square_d <- 0
    for (j in 1:10) {
      O_d <- sum(mix\$wid == i & mix\$did == j & mix\$ctime > start &
                 mix\$ctime < end)
      chi_square_d <- chi_square_d + (O_d - E_d)^2
    }
    chi_square_d <- chi_square_d / E_d
    p_d <- 1 - pchisq(chi_square_d, df=9)
    cat(sprintf("* Warehouse %d District chi-square p-value: %f\n", i, p_d))
  }
}
__EOF__
